src/cpp/core/DirectoryBasedFileLock.cpp

```diff
@@ -0,0 +1,222 @@
+/*
+ * DirectoryBasedFileLock.cpp
+ *
+ * Copyright (C) 2009-12 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/FileLock.hpp>
+
+#include <errno.h>
+#include <sys/stat.h>   // ::mkdir
+
+#include <set>
+#include <vector>
+
+#include <core/Thread.hpp>
+#include <core/Error.hpp>
+#include <core/Log.hpp>
+#include <core/FilePath.hpp>
+
+#include <boost/foreach.hpp>
+
+namespace rstudio {
+namespace core {
+
+namespace {
+
+std::string& rstudioLockingPath()
+{
+   static std::string instance("/tmp/rstudio-locking");
+   return instance;
+}
+
+class LockRegistration : boost::noncopyable
+{
+public:
+   
+   void registerLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.insert(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void deregisterLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.erase(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void refreshLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            lockFilePath.setLastWriteTime();
+         }
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void clearLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            lockFilePath.removeIfExists();
+         }
+      }
+      END_LOCK_MUTEX
+   }
+   
+   ~LockRegistration()
+   {
+      clearLocks();
+   }
+   
+private:
+   
+   boost::mutex mutex_;
+   std::set<FilePath> registration_;
+   
+};
+
+LockRegistration& lockRegistration()
+{
+   static LockRegistration instance;
```

Comment:
Let's confirm that the static destructor is actually called at process end on all platforms/compilers.